sdk/core/azure-core/inc/azure/core/uuid.hpp

```diff
@@ -54,6 +62,60 @@ namespace Azure { namespace Core {
      * @brief Construct a Uuid from an existing UUID represented as an array of bytes.
      * @details Creates a Uuid from a UUID created in an external scope.
      */
-    static Uuid CreateFromArray(std::array<uint8_t, UuidSize> const& uuid);
+    static constexpr Uuid CreateFromArray(ValueArray const& uuid) { return Uuid{uuid}; }
+
+    /**
+     * @brief Construct a Uuid by parsing its representation.
+     * @param s a string in `8-4-4-4-12` hex characters format.
+     * @throw `std::invalid_argument` if \p s cannot be parsed.
+     */
+    static Uuid Parse(std::string const& s);
+
+    /**
+     * @brief Compares with another instance of Uuid for equality.
+     * @param other another instance of Uuid.
+     * @return `true` if values of two Uuids are equal, `false` otherwise.
+     *
+     */
+    constexpr bool operator==(Uuid const& other) const
+    {
+      // std::array::operator==() is not a constexpr until C++20
+      for (size_t i = 0; i < m_uuid.size(); ++i)
+      {
+        if (m_uuid[i] != other.m_uuid[i])
+        {
+          return false;
+        }
+      }
+
+      return true;
+    }
+
+    /**
+     * @brief Compares with another instance of Uuid for inequality.
+     * @param other another instance of Uuid.
+     * @return `true` if values of two Uuids are not equal, `false` otherwise.
+     *
+     */
+    constexpr bool operator!=(Uuid const& other) const { return !(*this == other); }
+
+    /**
+     * @brief Checks if the value represents a Nil UUID (`00000000-0000-0000-0000-000000000000`).
+     *
+     */
+    constexpr bool IsNil() const
```

Comment:
In which scenario is this a useful helper?

Should we rather have a static 'Nil' or 'Zero' uuids constant, and then we don't need this helper method. Instead, the caller can use the equality operators for this check, instead.

Something like:
```c++
if (myUuid == Uuids::ZeroUuid)
{
// Take some action
}
```

This way, we don't need this additional method.

Reply:
Similar to .NET's "Empty"
https://learn.microsoft.com/en-us/dotnet/api/system.guid.empty?view=net-8.0#system-guid-empty

That is a more descriptive and unsurprising design, rather than the default ctor resulting in a "sentinel" nil value. I don't think having a default ctor on Uuid conveys the right intent.

Unless it is absolutely necessary, I would rather not have it in the public surface area. It doesn't have a "real" semantic meaning, on its own.

Reply:
I don't feel strongly about it bein either way, but Nil is an RFC's concept - not "Empty", not "Zero" - it is called Nil.
I think `IsNil()` is ok, it is similar to `HasValue()` for `Nullable`. It is possible to write a condition `if (!x.IsNil()) { /* serialize it */}`

`x == Uuid{}`, or `x == Uuid::Nil` are a bit more expensive. I don't think it will ever be significant though.

`x == Uuid{}` is a bit less readable, because you have to understand that `Uuid{}` produces Nil UUD.
But it is still possible to write it, if user wants it.

`Uuid::Nil` would require a DLLEXPORT, it is more expensive than `IsNil()`, it won't be a constexpr, while it also adds an API. So, we chose, which API to add, `IsNil()`, or `Uuid::Nil`, and I like `IsNil()` more.

Reply:
> `Uuid::Nil` would require a DLLEXPORT, it is more expensive than `IsNil()`

What do you mean by, it is more expensive?

Reply:
Compiler may treat it as a variable, not making an assumptions about its value at compile-time. Comparison with another array is a byte-by-byte comparison, `cmp *ptr, *ptr`, while `IsNil()` is `cmp *ptr, 0` or even `xor *ptr, *ptr`/`sub *ptr, *ptr`.